diff.js

```diff
@@ -466,21 +448,34 @@
           // Identical context lines. Track line changes
           if (oldRangeStart) {
             // Close out any changes that have been output (or join overlapping)
-            if (lines.length <= 8 && i < diff.length - 2) {
+            if (lines.length <= options.context * 2 && i < diff.length - 2) {
               // Overlapping
               curRange.push.apply(curRange, contextLines(lines));
             } else {
               // end the range and output
-              var contextSize = Math.min(lines.length, 4);
-              ret.push(
-                  '@@ -' + oldRangeStart + ',' + (oldLine - oldRangeStart + contextSize)
-                  + ' +' + newRangeStart + ',' + (newLine - newRangeStart + contextSize)
-                  + ' @@');
-              ret.push.apply(ret, curRange);
-              ret.push.apply(ret, contextLines(lines.slice(0, contextSize)));
-              if (lines.length <= 4) {
-                eofNL(ret, i, current);
+              var contextSize = Math.min(lines.length, options.context);
+              var hunklines = [];
+              hunklines.push.apply(hunklines, curRange);
+              hunklines.push.apply(hunklines, contextLines(lines.slice(0, contextSize)));
+
+              var hunk = { 
+                oldStart: oldRangeStart, 
+                oldLines: (oldLine - oldRangeStart + contextSize), 
+                newStart: newRangeStart, 
+                newLines: (newLine - newRangeStart + contextSize),
+                lines: hunklines
+              }
+              if(i >= diff.length - 2 && lines.length <= options.context) {
+                // EOF is inside this hunk
+                var oldEOFNewline = /\n$/.test(oldStr);
+                var newEOFNewline = /\n$/.test(newStr);
+                if(lines.length == 0 && !oldEOFNewline) {
+                  hunklines.splice(hunk.oldLines, 0, '\\ No newline at end of file')
+                } else if (!oldEOFNewline || !newEOFNewline) {
+                  hunklines.push('\\ No newline at end of file')
```

Comment:
lines.length == 0 && !oldEOFNewline is the special case where no-nl ends up in the middle of hunk. Will add comment to this effect.